### PR TITLE
Fix memory leaks

### LINKED ISSUE #1921: Memory leak
https://github.com/WebAssembly/binaryen/blob/master/src/passes/LegalizeJSInterface.cpp#L197

```C++
    auto* func = new Function;
    ...
    if (!module->getFunctionOrNull(func->name)) {
      // 1. internally it allocates unique_ptr for this pointer, so if this is executed, it is fine
      module->addFunction(func);  
    } // 2. however, if ELSE is executed, then it is a memory leak. 
```

Same with `legal` variable

### PR DESCRIPTION
**Pull request by Warchant**
Fixes #1921

### CODE REVIEW

**Comment by Warchant** [comment]
Please, re-run failed build https://travis-ci.org/WebAssembly/binaryen/jobs/499303663

File: CMakeLists.txt [hunk 1]
```diff
@@ -295,10 +297,10 @@ INSTALL(TARGETS wasm-reduce DESTINATION ${CMAKE_INSTALL_BINDIR})
...
+         COMMAND bin/wasm-opt ${CMAKE_SOURCE_DIR}/test/unit.wast --flatten --ssa --metrics -O4 -Os --metrics)
 ADD_TEST(NAME metrics-emcc
-         COMMAND bin/wasm-opt test/emcc_hello_world.fromasm --metrics)
+         COMMAND bin/wasm-opt ${CMAKE_SOURCE_DIR}/test/emcc_hello_world.fromasm --metrics)
 ADD_TEST(NAME exec-unit
-         COMMAND bin/wasm-opt test/unit.wast --fuzz-exec)
+         COMMAND bin/wasm-opt ${CMAKE_SOURCE_DIR}/test/unit.wast --fuzz-exec)
 ADD_TEST(NAME exec-hello
-         COMMAND bin/wasm-opt test/hello_world.wast --fuzz-exec)
+         COMMAND bin/wasm-opt ${CMAKE_SOURCE_DIR}/test/hello_world.wast --fuzz-exec)
```
Comment (sbc100) [inline CMakeLists.txt]: Can you make these cmakes a separate PR

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,6 +2,8 @@ PROJECT(binaryen C CXX)
 CMAKE_MINIMUM_REQUIRED(VERSION 2.8.7)
 INCLUDE(GNUInstallDirs)
 
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON) # generates compile_commands.json
```
Comment (sbc100) [inline CMakeLists.txt]: Is this for IDE integration?

File: CMakeLists.txt (hunk 1 above)
Comment (Warchant) [reply CMakeLists.txt]: Sure

File: CMakeLists.txt (hunk 2 above)
Comment (Warchant) [reply CMakeLists.txt]: This was necessary to run analyzers

**Comment by Warchant** [comment]
You can merge once CI succeeds

**Comment by kripken** [comment]
@Warchant - thanks! Have you joined the w3c group, before we merge? https://www.w3.org/community/webassembly/participants

**Comment by Warchant** [comment]
@kripken just signed in

File: src/passes/LegalizeJSInterface.cpp [hunk 3]
```diff
@@ -171,7 +171,7 @@ struct LegalizeJSInterface : public Pass {
     if (func->result == i64) {
       Function* f = getFunctionOrImport(module, SET_TEMP_RET0, "vi");
       legal->result = i32;
-      auto index = builder.addVar(legal, Name(), i64);
+      auto index = wasm::Builder::addVar(legal, Name(), i64);
```
Comment (kripken) [inline src/passes/LegalizeJSInterface.cpp]: no need for `wasm::` here

Comment (kripken) [reply src/passes/LegalizeJSInterface.cpp]: Looks good aside from this, thanks!
